ext/liquid_c/vm.c

```diff
@@ -0,0 +1,75 @@
+#include <stdint.h>
+
+#include "liquid.h"
+#include "vm.h"
+
+ID id_render_node;
+ID id_ivar_interrupts;
+ID id_ivar_resource_limits;
+ID id_ivar_render_length_limit;
+ID id_ivar_last_capture_length;
+ID id_increment_render_score;
+ID id_increment_write_score;
+ID id_raise_limits_reached;
+
+void liquid_vm_render(block_body_t *body, VALUE context, VALUE output)
+{
+    Check_Type(output, T_STRING);
+
+    VALUE resource_limits = rb_ivar_get(context, id_ivar_resource_limits);
+    rb_funcall(resource_limits, id_increment_render_score, 1, INT2NUM(body->render_score));
+
+    bool is_captured = rb_ivar_get(resource_limits, id_ivar_last_capture_length) != Qnil;;
+    long render_length_limit = LONG_MAX;
+
+    if (!is_captured) {
+        VALUE render_length_limit_num = rb_ivar_get(resource_limits, id_ivar_render_length_limit);
+        if (render_length_limit_num != Qnil)
+            render_length_limit = NUM2LONG(render_length_limit_num);
+    }
+
+    const size_t *const_ptr = (const size_t *)body->constants.data;
+    const uint8_t *ip = body->instructions.data;
+    VALUE interrupts = rb_ivar_get(context, id_ivar_interrupts);
+    Check_Type(interrupts, T_ARRAY);
+
+    while (true) {
```

Comment:
Would it make sense to do some tracking for how many bytes of the instruction buffer have been consumed, to gracefully handle a corrupted buffer with a missing OP_LEAVE? Perhaps by computing `last_ip` at the start and doing a single comparison per iteration.

Reply:
I would prefer to not silently try to handle a corrupt buffer, since that seems likely to lead to more problems.  However, I could add an assertion to help with debugging a problem like that.

Reply:
The idea would be to do something like an `rb_bug` in that case, yes.